// @progress/kendo-react-layout/dist/npm/bottomnavigation/BottomNavigationItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var kendo_licensing_1 = require("@progress/kendo-licensing");
var package_metadata_1 = require("../package-metadata");
var kendo_react_common_1 = require("@progress/kendo-react-common");
/**
 * Represents the [KendoReact BottomNavigationItem component]({% slug overview_bottomnavigation %}).
 */
exports.BottomNavigationItem = React.forwardRef(function (props, ref) {
    kendo_licensing_1.validatePackage(package_metadata_1.packageMetadata);
    var elementRef = React.useRef(null);
    React.useImperativeHandle(ref, function () { return ({
        element: elementRef.current
    }); });
    var className = props.className, style = props.style, selected = props.selected, disabled = props.disabled, item = props.item, render = props.render, dataItem = props.dataItem, icon = props.icon, text = props.text, id = props.id, onSelect = props.onSelect, onKeyDown = props.onKeyDown, index = props.index, tabIndex = props.tabIndex;
    var itemClasses = React.useMemo(function () { return kendo_react_common_1.classNames('k-bottom-nav-item', {
        'k-state-selected': selected,
        'k-state-disabled': disabled
    }, className); }, [selected, disabled, className]);
    var handleClick = React.useCallback(function (event) {
        if (onSelect && index !== undefined && !disabled) {
            onSelect(event, index);
        }
    }, [onSelect, index, disabled]);
    var handleKeyDown = React.useCallback(function (event) {
        if (onKeyDown && index !== undefined && !disabled) {
            onKeyDown(event, index);
        }
    }, [onKeyDown, index, disabled]);
    var renderer = render;
    var ItemComp = item;
    var bottomNavItem = (React.createElement("span", { ref: elementRef, className: itemClasses, style: style, role: 'link', id: id, tabIndex: tabIndex, onClick: handleClick, onKeyDown: handleKeyDown, "aria-selected": selected, "aria-disabled": disabled }, ItemComp ? React.createElement(ItemComp, { itemIndex: index, item: dataItem }) :
        (React.createElement(React.Fragment, null,
            icon && React.createElement("span", { className: "k-bottom-nav-item-icon k-icon k-i-" + icon }),
            text && React.createElement("span", { className: 'k-bottom-nav-item-text', style: { userSelect: 'none' } }, text)))));
    return renderer !== undefined ? renderer.call(undefined, bottomNavItem, props) : bottomNavItem;
});
var defaultProps = {
    tabIndex: 0
};
exports.BottomNavigationItem.propTypes = {
    className: PropTypes.string,
    style: PropTypes.object,
    id: PropTypes.string,
    disabled: PropTypes.bool,
    selected: PropTypes.bool,
    icon: PropTypes.string,
    text: PropTypes.string,
    tabIndex: PropTypes.number
};
exports.BottomNavigationItem.defaultProps = defaultProps;
exports.BottomNavigationItem.displayName = 'KendoReactBottomNavigationItem';
