// @progress/kendo-react-layout/dist/npm/bottomnavigation/BottomNavigation.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var kendo_licensing_1 = require("@progress/kendo-licensing");
var package_metadata_1 = require("../package-metadata");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var utils_1 = require("./models/utils");
var BottomNavigationItem_1 = require("./BottomNavigationItem");
/**
 * Represents the [KendoReact BottomNavigation component]({% slug overview_bottomnavigation %}).
 * Used to switch between primary destinations in an application.
 *
 * @example
 * ```jsx
 * const items = [
 *    { text: 'Photos', icon: 'photo', selected: true },
 *    { text: 'Albums', icon: 'folder' },
 *    { text: 'Search', icon: 'search' },
 *    { text: 'More', icon: 'more-horizontal' },
 * ];
 *
 * const App = () => {
 *   return (
 *       <BottomNavigation items={items} />
 *   );
 * };
 *    ReactDOM.render(<App />, document.querySelector('my-app'));
 * ```
 */
exports.BottomNavigation = React.forwardRef(function (props, ref) {
    kendo_licensing_1.validatePackage(package_metadata_1.packageMetadata);
    var elementRef = React.useRef(null);
    var focusElement = React.useCallback(function () {
        if (elementRef.current) {
            kendo_react_common_1.focusFirstFocusableChild(elementRef.current);
        }
    }, []);
    var getImperativeHandle = React.useCallback(function () { return ({
        element: elementRef.current,
        focus: focusElement
    }); }, [focusElement]);
    React.useImperativeHandle(ref, getImperativeHandle);
    var _a = props.themeColor, themeColor = _a === void 0 ? defaultProps.themeColor : _a, _b = props.fill, fill = _b === void 0 ? defaultProps.fill : _b, _c = props.positionMode, positionMode = _c === void 0 ? defaultProps.positionMode : _c, _d = props.itemFlow, itemFlow = _d === void 0 ? defaultProps.itemFlow : _d, _e = props.border, border = _e === void 0 ? defaultProps.border : _e, className = props.className, items = props.items, item = props.item, itemRender = props.itemRender, disabled = props.disabled, style = props.style, id = props.id, onSelect = props.onSelect, onKeyDown = props.onKeyDown;
    var navId = React.useMemo(function () { return kendo_react_common_1.guid(); }, []);
    var dir = kendo_react_common_1.useDir(elementRef, props.dir);
    var navClasses = React.useMemo(function () { return kendo_react_common_1.classNames('k-bottom-nav', utils_1.POSITION_MODE_CLASSES[positionMode], utils_1.THEME_COLOR_CLASSES[themeColor], utils_1.FILL_CLASSES[fill], utils_1.ITEM_FLOW_CLASSES[itemFlow], {
        'k-bottom-nav-border': border,
        'k-state-disabled': disabled
    }, className); }, [positionMode, themeColor, fill, itemFlow, border, disabled, className]);
    var dispatchSelectEvent = React.useCallback(function (dispatchedEvent, index) {
        if (!items) {
            return;
        }
        if (!items[index].disabled && onSelect) {
            kendo_react_common_1.dispatchEvent(onSelect, dispatchedEvent, getImperativeHandle(), {
                itemTarget: items[index],
                itemIndex: index
            });
        }
    }, [items, onSelect]);
    var handleSelect = React.useCallback(function (event, clickedItemIndex) {
        dispatchSelectEvent(event, clickedItemIndex);
    }, [dispatchSelectEvent]);
    var handleKeyDown = React.useCallback(function (event, clickedItemIndex) {
        switch (event.keyCode) {
            case kendo_react_common_1.Keys.enter:
            case kendo_react_common_1.Keys.space:
                dispatchSelectEvent(event, clickedItemIndex);
                event.preventDefault();
                break;
            default:
                break;
        }
        kendo_react_common_1.dispatchEvent(onKeyDown, event, getImperativeHandle(), undefined);
    }, [dispatchSelectEvent, onKeyDown]);
    return (React.createElement("nav", { ref: elementRef, className: navClasses, style: style, id: id || navId, dir: dir }, items && (items.map(function (element, index) {
        return (React.createElement(BottomNavigationItem_1.BottomNavigationItem, __assign({}, element, { key: index, index: index, id: (id || navId) + "-" + index, disabled: disabled || element.disabled, selected: element.selected, dataItem: element, item: item, render: itemRender, onSelect: handleSelect, onKeyDown: handleKeyDown })));
    }))));
});
var defaultProps = {
    themeColor: 'primary',
    fill: 'flat',
    itemFlow: 'vertical',
    positionMode: 'fixed',
    border: true
};
exports.BottomNavigation.propTypes = {
    className: PropTypes.string,
    style: PropTypes.object,
    id: PropTypes.string,
    dir: PropTypes.string,
    themeColor: PropTypes.oneOf([
        'primary', 'secondary', 'tertiary',
        'info', 'success', 'warning',
        'error', 'dark', 'light', 'inverse'
    ]),
    fill: PropTypes.oneOf(['solid', 'flat']),
    itemFlow: PropTypes.oneOf(['vertical', 'horizontal']),
    border: PropTypes.bool,
    disabled: PropTypes.bool,
    selected: PropTypes.number,
    onSelect: PropTypes.func
};
exports.BottomNavigation.defaultProps = defaultProps;
exports.BottomNavigation.displayName = 'KendoReactBottomNavigation';
